# **Data Loading**

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### **Data set yang digunakan berasal dari Kaggle [Multiple Disease Prediction](https://www.kaggle.com/datasets/ehababoelnaga/multiple-disease-prediction/data)**

In [2]:
import zipfile
import os

zip_file_path = '/content/drive/MyDrive/Multiple Disease Prediction.zip'

# Ekstrak isi file zip
extract_dir = '/content/multiple-disease-prediction'
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_dir)

# List semua file yang diekstrak
files = os.listdir(extract_dir)

# Cetak daftar nama file
for file_name in files:
    print(file_name)


Blood_samples_dataset_balanced_2(f).csv
blood_samples_dataset_test.csv


# **Import Library**

In [3]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

# **Exploratory Data**

In [4]:
df_train = pd.read_csv('/content/multiple-disease-prediction/Blood_samples_dataset_balanced_2(f).csv')
df_test = pd.read_csv('/content/multiple-disease-prediction/blood_samples_dataset_test.csv')

## **Train Data**

In [5]:
df_train.head()

Glucose  Cholesterol  Hemoglobin  Platelets  White Blood Cells  \
0  0.739597     0.650198    0.713631   0.868491           0.687433   
1  0.121786     0.023058    0.944893   0.905372           0.507711   
2  0.452539     0.116135    0.544560   0.400640           0.294538   
3  0.136609     0.015605    0.419957   0.191487           0.081168   
4  0.176737     0.752220    0.971779   0.785286           0.443880   

   Red Blood Cells  Hematocrit  Mean Corpuscular Volume  \
0         0.529895    0.290006                 0.631045   
1         0.403033    0.164216                 0.307553   
2         0.382021    0.625267                 0.295122   
3         0.166214    0.073293                 0.668719   
4         0.439851    0.894991                 0.442159   

   Mean Corpuscular Hemoglobin  Mean Corpuscular Hemoglobin Concentration  \
0                     0.001328                                   0.795829   
1                     0.207938                                   0.505562   
2                     0.868369                                   0.026808   
3                     0.125447                                   0.501051   
4                     0.257288                                   0.805987   

   ...     HbA1c  LDL Cholesterol  HDL Cholesterol       ALT       AST  \
0  ...  0.502665         0.215560         0.512941  0.064187  0.610827   
1  ...  0.856810         0.652465         0.106961  0.942549  0.344261   
2  ...  0.466795         0.387332         0.421763  0.007186  0.506918   
3  ...  0.016256         0.040137         0.826721  0.265415  0.594148   
4  ...  0.429431         0.146294         0.221574  0.015280  0.567115   

   Heart Rate  Creatinine  Troponin  C-reactive Protein   Disease  
0    0.939485    0.095512  0.465957            0.769230   Healthy  
1    0.666368    0.659060  0.816982            0.401166  Diabetes  
2    0.431704    0.417295  0.799074            0.779208  Thalasse  
3    0.225756    0.490349  0.637061            0.354094    Anemia  
4    0.841412    0.153350  0.794008            0.094970  Thalasse  

[5 rows x 25 columns]

In [6]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2351 entries, 0 to 2350
Data columns (total 25 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Glucose                                    2351 non-null   float64
 1   Cholesterol                                2351 non-null   float64
 2   Hemoglobin                                 2351 non-null   float64
 3   Platelets                                  2351 non-null   float64
 4   White Blood Cells                          2351 non-null   float64
 5   Red Blood Cells                            2351 non-null   float64
 6   Hematocrit                                 2351 non-null   float64
 7   Mean Corpuscular Volume                    2351 non-null   float64
 8   Mean Corpuscular Hemoglobin                2351 non-null   float64
 9   Mean Corpuscular Hemoglobin Concentration  2351 non-null   float64
 10  Insulin                 

In [7]:
df_train.isna().sum()

Glucose                                      0
Cholesterol                                  0
Hemoglobin                                   0
Platelets                                    0
White Blood Cells                            0
Red Blood Cells                              0
Hematocrit                                   0
Mean Corpuscular Volume                      0
Mean Corpuscular Hemoglobin                  0
Mean Corpuscular Hemoglobin Concentration    0
Insulin                                      0
BMI                                          0
Systolic Blood Pressure                      0
Diastolic Blood Pressure                     0
Triglycerides                                0
HbA1c                                        0
LDL Cholesterol                              0
HDL Cholesterol                              0
ALT                                          0
AST                                          0
Heart Rate                                   0
Creatinine   

## **Test Data**

In [8]:
df_test.head()

Glucose  Cholesterol  Hemoglobin  Platelets  White Blood Cells  \
0  0.001827     0.033693    0.114755   0.997927           0.562604   
1  0.436679     0.972653    0.084998   0.180909           0.675736   
2  0.545697     0.324815    0.584467   0.475748           0.558596   
3  0.172994     0.050351    0.736000   0.782022           0.069435   
4  0.758534     0.739968    0.597868   0.772683           0.875720   

   Red Blood Cells  Hematocrit  Mean Corpuscular Volume  \
0         0.866499    0.578042                 0.914615   
1         0.563889    0.798382                 0.670361   
2         0.661007    0.934056                 0.381782   
3         0.085219    0.032907                 0.460619   
4         0.860265    0.486189                 0.486686   

   Mean Corpuscular Hemoglobin  Mean Corpuscular Hemoglobin Concentration  \
0                     0.026864                                   0.038641   
1                     0.376092                                   0.184890   
2                     0.500342                                   0.531829   
3                     0.785448                                   0.491495   
4                     0.621048                                   0.191756   

   ...     HbA1c  LDL Cholesterol  HDL Cholesterol       ALT       AST  \
0  ...  0.653230         0.186104         0.430398  0.016678  0.885352   
1  ...  0.833540         0.153001         0.458533  0.401845  0.635969   
2  ...  0.678901         0.220479         0.817151  0.690981  0.101633   
3  ...  0.381500         0.459396         0.420154  0.798537  0.399236   
4  ...  0.993381         0.272338         0.663579  0.265227  0.918847   

   Heart Rate  Creatinine  Troponin  C-reactive Protein   Disease  
0    0.652733    0.788235  0.054788            0.031313  Thalasse  
1    0.574425    0.047025  0.607985            0.594123  Diabetes  
2    0.855740    0.551124  0.413294            0.070909  Heart Di  
3    0.324600    0.499504  0.436662            0.242766  Diabetes  
4    0.804910    0.571119  0.188368            0.750848  Heart Di  

[5 rows x 25 columns]

In [9]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 486 entries, 0 to 485
Data columns (total 25 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Glucose                                    486 non-null    float64
 1   Cholesterol                                486 non-null    float64
 2   Hemoglobin                                 486 non-null    float64
 3   Platelets                                  486 non-null    float64
 4   White Blood Cells                          486 non-null    float64
 5   Red Blood Cells                            486 non-null    float64
 6   Hematocrit                                 486 non-null    float64
 7   Mean Corpuscular Volume                    486 non-null    float64
 8   Mean Corpuscular Hemoglobin                486 non-null    float64
 9   Mean Corpuscular Hemoglobin Concentration  486 non-null    float64
 10  Insulin                   

In [10]:
df_test.isna().sum()

Glucose                                      0
Cholesterol                                  0
Hemoglobin                                   0
Platelets                                    0
White Blood Cells                            0
Red Blood Cells                              0
Hematocrit                                   0
Mean Corpuscular Volume                      0
Mean Corpuscular Hemoglobin                  0
Mean Corpuscular Hemoglobin Concentration    0
Insulin                                      0
BMI                                          0
Systolic Blood Pressure                      0
Diastolic Blood Pressure                     0
Triglycerides                                0
HbA1c                                        0
LDL Cholesterol                              0
HDL Cholesterol                              0
ALT                                          0
AST                                          0
Heart Rate                                   0
Creatinine   

# **Data Preparation**

In [11]:
# Menggabungkan dataset df_train dan df_test

df = pd.concat([df_train, df_test], ignore_index=True)

In [12]:
X = df.drop(columns=['Disease']) # berisi fitur-fitur dari data tanpa kolom 'Disease'
y = df['Disease'] # berisi kolom 'Disease' sebagai target atau label yang akan diprediksi

In [13]:
# Membagi dataset menjadi data pelatihan dan test

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)


# **Model Development**

In [14]:
# Model Logistic Regression

lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [15]:
# Model Random Forest

rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)

RandomForestClassifier()

In [16]:
# Model SVM

svm_model = SVC()
svm_model.fit(X_train, y_train)

SVC()

# **Evaluasi Model**

In [17]:
results = []

models = [
    ('Logistic Regression', lr_model.predict(X_test)),
    ('Random Forest', rf_model.predict(X_test)),
    ('SVM', svm_model.predict(X_test))
]

for model_name, y_pred in models:
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')

    results.append({
        'Model': model_name,
        'Accuracy': accuracy,
        'Precision': precision,
        'Recall': recall,
        'F1 Score': f1
    })

df_results = pd.DataFrame(results)

print(df_results)


                 Model  Accuracy  Precision    Recall  F1 Score
0  Logistic Regression  0.797535   0.791900  0.797535  0.790412
1        Random Forest  0.963028   0.953073  0.963028  0.956784
2                  SVM  0.950704   0.939362  0.950704  0.944295


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Berdasarkan hasil pelatihan model yang dilakukan, dapat disimpulkan bahwa:

*   Model Logistik Regresi memiliki tingkat akurasi sebesar 79.75%, dengan nilai presisi sebesar 79.19%, recall sebesar 79.75%, dan nilai F1 sebesar 79.04%
*   Model Random Forest menunjukkan tingkat akurasi tertinggi, mencapai 95.95%, dengan presisi sebesar 94.95%, recall sebesar 95.95%, dan nilai F1 sebesar 95.32%.
*   Model Support Vector Machine (SVM) juga menunjukkan kinerja yang baik dengan tingkat akurasi sebesar 95.07%, nilai presisi sebesar 93.94%, recall sebesar 95.07%, dan nilai F1 sebesar 94.43%.

Dari kesimpulan tersebut, dapat disimpulkan bahwa Model Random Forest memiliki kinerja terbaik dalam hal akurasi dibandingkan dengan model Logistik Regresi dan SVM.